File: solutions/javascript/lasagna-master/1/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Determine whether the lasagna is done
 */
export function cookingStatus(remainingTime) {
    let status
    if ( remainingTime === undefined ) {
        status = 'You forgot to set the timer.'
    } else if ( remainingTime === 0 ) {
        status = 'Lasagna is done.'
    } else {
        status = 'Not done, please wait.'
    }
    return status
}

/**
 * Estimate the preparation time
 */
export function preparationTime(
    layers = [], averagePreparationTimePerLayer = 2
) {
    return layers.length * averagePreparationTimePerLayer
}

/**
 * Compute the amounts of noodles and sauce needed
 */
export function quantities(layers = []) {
    let adds = {
        noodles: 50,
        sauce: 0.2,
    }
    let amounts = {
        noodles: 0,
        sauce:   0,
    }
    for ( let i in layers ) {
        if ( adds.hasOwnProperty(layers[i]) ) {
            amounts[layers[i]] += adds[layers[i]]
        }
    }
    return amounts
}

/**
 * Add the secret ingredient
 */
export function addSecretIngredient(
    friendIngredients = [], ownIngredients = []
) {
    ownIngredients.push(friendIngredients[friendIngredients.length - 1])
}

/**
 * Scale the recipe
 */
export function scaleRecipe(recipeForTwoPortions, numberOfPortions) {
    let recipe     = {}
    let multiplier = numberOfPortions / 2
    for ( let item in recipeForTwoPortions ) {
        recipe[item] = recipeForTwoPortions[item] * multiplier;
    }
    return recipe
}

File: solutions/go/lasagna-master/1/lasagna_master.go
package lasagna

// define the 'PreparationTime()' function
func PreparationTime(layers []string, averageMinutes int) (prepTime int) {
    if averageMinutes == 0 {
        averageMinutes = 2
    }
    prepTime = len(layers) * averageMinutes
    return
}

// define the 'Quantities()' function
func Quantities(layers []string) (noodles int, sauce float64) {
    noodles = 0
    sauce   = 0
    for i := 0; i < len(layers); i++ {
        switch layers[i] {
        case "noodles": noodles += 1
        case "sauce"  :   sauce += 1.0
        }
    }
    noodles *= 50
    sauce   *=  0.2
    return
}

// define the 'AddSecretIngredient()' function
func AddSecretIngredient(frindList, myList []string) {
    secretIngredient := frindList[len(frindList) - 1]
    myList[len(myList) - 1] = secretIngredient
}

// define the 'ScaleRecipe()' function
func ScaleRecipe(quantities []float64, portions int) []float64 {
    amounts := make([]float64, len(quantities))
    scale := float64(portions) / 2.0
    for i := 0; i < len(quantities); i++ {
        amounts[i] = quantities[i] * scale
    }
    return amounts
}
